pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,856 @@
+                   "'value'. Got a '{}' instead of a 'pandas.Interval'."
+                   .format(type(value).__name__))
+            raise TypeError(msg)
+
+        value = getattr(value, '_values', value)
+
+        left = self.left.fillna(value=value.left)
+        right = self.right.fillna(value=value.right)
+        return self._shallow_copy(left, right)
+
+    @property
+    def dtype(self):
+        return IntervalDtype.construct_from_string(str(self.left.dtype))
+
+    def astype(self, dtype, copy=True):
+        dtype = pandas_dtype(dtype)
+        if is_interval_dtype(dtype) and dtype != self.dtype:
+            try:
+                new_left = self.left.astype(dtype.subtype)
+                new_right = self.right.astype(dtype.subtype)
+            except TypeError:
+                msg = ('Cannot convert {dtype} to {new_dtype}; subtypes are '
+                       'incompatible')
+                raise TypeError(msg.format(dtype=self.dtype, new_dtype=dtype))
+            return self._shallow_copy(new_left, new_right)
+        elif is_interval_dtype(dtype):
+            if copy:
+                return self.copy()
+            else:
+                return self
+        elif is_categorical_dtype(dtype):
+            return Categorical(self.values)
+        # TODO: This try/except will be repeated.
+        try:
+            return self.values.astype(dtype, copy=copy)
+        except (TypeError, ValueError):
+            msg = 'Cannot cast {name} to dtype {dtype}'
+            raise TypeError(msg.format(name=type(self).__name__, dtype=dtype))
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        closed = set(interval.closed for interval in to_concat)
+        if len(closed) != 1:
+            raise ValueError("Intervals must all be closed on the same side.")
+        closed = closed.pop()
+
+        # TODO: avoid intermediate list
+        left = np.concatenate([interval.left for interval in to_concat])
+        right = np.concatenate([interval.right for interval in to_concat])
+        return cls._simple_new(left, right, closed=closed, copy=False)
+
+    def _shallow_copy(self, left=None, right=None):
+        from pandas.core.indexes.interval import IntervalIndex
+
+        if left is None:
+
+            # no values passed
+            # XXX: is ^ right? Or does that mean just left wasn't passed?
+            left, right = self.left, self.right
+
+        elif right is None:
+
+            # only single value passed, could be an IntervalIndex
+            # or array of Intervals
+            if not isinstance(left, (type(self), IntervalIndex)):
+                left = type(self)(left)
+
+            left, right = left.left, left.right
+        else:
+
+            # both left and right are values
+            pass
+
+        return self._simple_new(left, right, closed=self.closed,
+                                verify_integrity=False)
+
+    # TODO: doc
+    def copy(self, deep=False):
```

Comment:
is this in EA?

Reply:
Yes, `copy` is part of the extension array interface